cmd/kube-apiserver/app/server.go

```diff
@@ -412,7 +412,7 @@ func buildGenericConfig(
 		lastErr = err
 		return
 	}
-	storageFactory, lastErr = completedStorageFactoryConfig.New()
+	storageFactory, lastErr = completedStorageFactoryConfig.New(genericConfig.DrainedNotify())
```

Comment:
My small concern with this is that draining currently doesn't drain long-running requests, in particular watches.

However, watches can also require KMS or decrypting stuff they receive from etcd.

It's technically not end of the world, because it will just result in sending an error event:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L331
and watchers should recover from it.
But in some cases, it may cause unnecessary relists:
https://github.com/kubernetes/kubernetes/blob/09e8339ae4e575ec40e8ed0fc2829fcb678c0f19/staging/src/k8s.io/client-go/tools/cache/reflector.go#L473
[FWIW - I wanted to address it for some time, but never got to it...]

I'm wondering if we shouldn't use some other signal, like HTTPServerStoppedListening in the meantime.

Reply:
I'm not familiar with this, but it seems the apiserver stops to receive request a few steps earlier

https://github.com/kubernetes/kubernetes/blob/253acfc8bce430ed835f14d5ec062356701cff2f/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L417-L458

has to be this the last thing stopping?

Reply:
put another way, it sounds like using HTTPServerStoppedListening means in-flight write requests will now fail, not just in-flight long-lived watch requests

Reply:
I'm not sure I follow, listening vs accepting has different meanings 

https://github.com/kubernetes/kubernetes/blob/9644edc321e3b7e5180eb7c15a27bf28e19485db/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L563-L564

Reply:
> that would seem to indicate we should use InFlightRequestsDrained not HTTPServerStoppedListening, right?

As I mentioned above, InFlightRequestsDrained does NOT include long-running requests. So that's not perfect.
@liggitt - I think the comment you pasted, is not really up-to-date.

I would appreciate if someone would double check me, but the  HTTPServerStoppedListening channel:
https://github.com/kubernetes/kubernetes/blob/253acfc8bce430ed835f14d5ec062356701cff2f/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L539

is effectively closed here:
https://github.com/kubernetes/kubernetes/blob/253acfc8bce430ed835f14d5ec062356701cff2f/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go#L242
So when http server is already down. Which basically happens after the requests are drained.


I agree that HTTPServerStoppedListening is not a perfect signal [we're probably missing one more signal], but I still think it's better than the Drained one which doesn't include long-running requests.

Reply:
After reading the code I believe:

Neither `InFlightRequestsDrained` nor `HTTPServerStoppedListening` handle long running requests.  We have no option to handle that case AFAIK.

>  So when http server is already down. Which basically happens after the requests are drained.
> 
> I agree that HTTPServerStoppedListening is not a perfect signal [we're probably missing one more signal], but I still think it's better than the Drained one which doesn't include long-running requests.

This is not true.  `HTTPServerStoppedListening` corresponds to the listener socket being closed.  It does not close connections that have already been accepted (i.e. watches) by that socket (they have distinct FDs).

Thus I believe `InFlightRequestsDrained` is the best option available to us because:

1. We know new connections will not be made
2. We know that we have attempted to drain non-long running requests

Reply:
@wojtek-t @liggitt ping re ^

Reply:
> Thus I believe InFlightRequestsDrained is the best option available to us because:
>
> 1. We know new connections will not be made
> 2. We know that we have attempted to drain non-long running requests

that makes sense to me

Reply:
@liggitt this is ready now.